src/main/java/net/neoforged/neoforge/transfer/resources/ResourceStack.java

```diff
@@ -0,0 +1,108 @@
+            amount = 0;
+        } else {
+            // if the resource is not empty, but was passed in 0 for amount, this is invalid. This is to minimize misuse.
+            if (amount == 0) {
+                //We have to throw as we don't have access to the empty instance, nor can we defer to an instanced stack with a constructor.
+                //This info would need to be provided by the resource in some way which we currently don't have.
+                throw new IllegalArgumentException("An empty resource must be used when amount is 0; " + resource + "(0) is not allowed");
+            }
+        }
+    }
+
+    /**
+     * Creates a codec with the resource being a field in the object. Should the resource or amount be empty, the empty resource stack instance
+     * provided by the stack factory will be used.
+     *
+     * <pre>{@code
+     * {
+     *     "resource": {
+     *         "id": "minecraft:water",
+     *         "components": { ... }
+     *     },
+     *     "amount": 1000
+     * }
+     * }</pre>
+     *
+     * @param resourceCodec a codec for the resource
+     * @param stackFactory  The stack factory handling new instance creation or deferring to the empty stack instance.
+     * @param <R>           the resource type
+     * @return a codec for a resource stack
+     */
+    public static <R extends IResource> Codec<ResourceStack<R>> codec(Codec<R> resourceCodec, IStackFactory<R, ResourceStack<R>> stackFactory) {
+        return RecordCodecBuilder.create(instance -> instance.group(
+                resourceCodec.fieldOf("resource").forGetter(ResourceStack::resource),
+                NeoForgeExtraCodecs.optionalFieldAlwaysWrite(ExtraCodecs.NON_NEGATIVE_INT, "amount", 1).forGetter(ResourceStack::amount))
+                .apply(instance, stackFactory::create));
+    }
+
+    /**
+     * Creates a standard stream codec for a ResourceStack of the specified resource type. Should the resource or amount be empty, the empty resource stack instance
+     * provided by the resource will be used.
+     *
+     * @param resourceCodec The codec for the resource type.
+     * @param stackFactory  The stack factory handling new instance creation or deferring to the empty stack instance.
+     * @param <R>           the resource type
+     */
+    public static <B extends FriendlyByteBuf, R extends IResource> StreamCodec<B, ResourceStack<R>> streamCodec(StreamCodec<? super B, R> resourceCodec, IStackFactory<R, ResourceStack<R>> stackFactory) {
+        return StreamCodec.composite(
+                resourceCodec, ResourceStack::resource,
+                ByteBufCodecs.VAR_INT, ResourceStack::amount,
+                stackFactory::create);
+    }
+
+    /**
+     * Checks if the resource stack is empty, meaning that the amount is zero
+     * or that the resource is {@link IResource#isEmpty() empty}.
+     *
+     * @return {@code true} if empty
+     */
+    public boolean isEmpty() {
```

Comment:
This might be okay to stay as a helper, personally indifferent, other maintainers can weigh in here.